### PR TITLE
Introduce decimal type in sql

### PR DESCRIPTION
**Pull request by ImeevMA**
First 8 commits from another patch-set.

After this patch-set decimal will be available in SQL. It will have defined
comparison operations, arithmetic operations, implicit and explicit casts.
However, some functions cannot work with DECIMAL for now, and it will be fixed a
bit later in patch-set that fixes SQL built-in functions. Also, for now DECIMAL
will not have literals in SQL, since it still was not decided how to implement
them.

### CODE REVIEW

**Review by Korablev77 — COMMENTED** [review]
sql: remove enum field_type from struct Mem

I don't support the decision to make this patch, but in terms of code quality - it is OK.

**Review by Korablev77 — COMMENTED** [review]
2) Any value of any scalar type can be implicitly cast to SCALAR;
...
4) SCALAR values cannot be implicitly cast to any other scalar type;

Don't get it. Aren't these points contradictory?

File: src/box/sql/mem.c [hunk 1]
```diff
@@ -1733,6 +1733,18 @@ mem_get_int(const struct Mem *mem, int64_t *i, bool *is_neg)
 		}
 		return -1;
 	}
+	if (mem->type == MEM_TYPE_DEC) {
+		if (decimal_is_neg(&mem->u.d)) {
+			if (decimal_to_int64(&mem->u.d, i) == NULL)
+				return -1;
+			*is_neg = *i < 0;
```
Comment (Korablev77) [inline src/box/sql/mem.c:1740]: If decimal is negative then how integer can be positive? In case it is zero check I guess it's better write (*i == 0)

**Comment by Korablev77** [comment]
Actually I do not like the mess in the patch-set. IMHO series should contain only changes related to decimal part, meanwhile patches in it touches different unrelated problems...

File: test/sql-tap/decimal.test.lua [hunk 2]
```diff
@@ -425,6 +425,370 @@ test:do_catchsql_test(
...
+    ]], {
+        dec1, dec3, dec2
+    })
+
+test:do_execsql_test(
+    "dec-10.1.4",
+    [[
+        SELECT cast(u AS DOUBLE) FROM t2;
+    ]], {
+        111, 3333, 55555
```
Comment (Korablev77) [inline test/sql-tap/decimal.test.lua:458]: Why not 111.0, 3333.0, 55555.0?

File: src/box/sql/mem.c [hunk 3]
```diff
@@ -1590,23 +1590,31 @@ mem_concat(struct Mem *a, struct Mem *b, struct Mem *result)
 	return 0;
 }
 
+static inline int
+check_types_numeric_arithmetic(const struct Mem *a, const struct Mem *b)
```
Comment (Korablev77) [inline src/box/sql/mem.c:2004]: mem_are_arithmetic_compatible() sounds better.
Or mem_types_are_arithmetical()

**Comment by pgulutzan** [comment]
Like @Korablev77, I don't support the decision to make this patch.
But I downloaded imeevma-gh-4415-introduce-decimal-type-in-sql anyway.

@tsafin wrote in https://github.com/tarantool/tarantool/discussions/5910
"Rework the SQL parser so that all non-integral numeric literals have
DECIMAL type with corresponding precision and scale."
Now I see that, as I expected, it "still was not decided".
Most of the things that I hoped for in #4415 are not there ...
The abbreviation DEC is not okay.
DECIMAL is not the same as DECIMAL(38,0) because DECIMAL(38,0) is illegal.
SUM does not work.
Scalar functions do not work.
And this causes a segmentation fault:
box.execute([[CREATE TABLE t (s1 DECIMAL PRIMARY KEY) WITH ENGINE='vinyl';]])
box.execute([[INSERT INTO t VALUES (5.5);]])

File: src/box/sql/mem.c [hunk 4]
```diff
@@ -988,6 +1023,69 @@ double_to_uint_forced(struct Mem *mem)
 	return res;
 }
 
+static inline int
+double_to_dec(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DOUBLE);
+	double d = mem->u.r;
+	if (d >= 1e38 || d <= -1e38)
```
Comment (Korablev77) [inline src/box/sql/mem.c:1031]: Why do you need this additional check?

Comment (ImeevMA) [reply src/box/sql/mem.c:1031]: Because cast of values more than 1e38 and less than -1e38 are out of DECIMAL range. So CAST should fail.

**Comment by ImeevMA** [comment]
@pgulutzan Thank you for your comments!

> Like @Korablev77, I don't support the decision to make this patch.
> But I downloaded imeevma-[gh-4415](https://github.com/tarantool/tarantool/issues/4415)-introduce-decimal-type-in-sql anyway.
> 
> @tsafin wrote in #5910
> "Rework the SQL parser so that all non-integral numeric literals have
> DECIMAL type with corresponding precision and scale."
> Now I see that, as I expected, it "still was not decided".
That is quite problematic in current parser design, since we have to determine of value is INTEGER or DECIMAL. Since we not always can see sign of the value, we have some problems due to the fact that INTEGER is not symmetrical.

> Most of the things that I hoped for in #4415 are not there ...
> The abbreviation DEC is not okay.
I will add this.

> DECIMAL is not the same as DECIMAL(38,0) because DECIMAL(38,0) is illegal.
We have no such modifiers for now.

> SUM does not work.
I mentioned that built-in functions will be in another patch in this release.

> Scalar functions do not work.
Same.

> And this causes a segmentation fault:
> box.execute([[CREATE TABLE t (s1 DECIMAL PRIMARY KEY) WITH ENGINE='vinyl';]])
> box.execute([[INSERT INTO t VALUES (5.5);]])
This is my bug. I already found where is the error and how to fix it.

File: src/box/sql/mem.c (hunk 4 above)
Comment (ImeevMA) [reply src/box/sql/mem.c:1031]: Or do you mean that we just have to see result of `decimal_from_double()`?

Comment (Korablev77) [reply src/box/sql/mem.c:1031]: > Or do you mean that we just have to see result of decimal_from_double()?

Yep

File: src/box/sql/mem.c [hunk 5]
```diff
@@ -1001,6 +1099,173 @@ double_to_str0(struct Mem *mem)
...
+			mem->u.u = 0;
+			return -1;
+		}
+		mem->u.u = UINT64_MAX;
+		return 1;
+	}
+	mem->u.u = u;
+	/*
+	 * Decimal is floored when cast to uint, which means that after cast if
+	 * if was not integer it becomes less if it was positive, and move if it
```
Comment (Korablev77) [inline src/box/sql/mem.c:1209]: if -> it
Please add example to comments, it turns out to be hard to parse...

File: src/box/sql/mem.c [hunk 6]
```diff
@@ -988,6 +1023,69 @@ double_to_uint_forced(struct Mem *mem)
...
+double_to_dec_forced(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DOUBLE);
+	double d = mem->u.r;
+	mem->type = MEM_TYPE_DEC;
+	mem->flags = 0;
+	if (d >= 1e38) {
+		const char *val = "99999999999999999999999999999999999999";
+		assert(strlen(val) == 38);
+		decimal_from_string(&mem->u.d, val);
```
Comment (Korablev77) [inline src/box/sql/mem.c:1071]: These conversions from string looks quite strange. Could we expand somehow decimal API and add direct method to achieve this?

File: test/sql-tap/decimal.test.lua [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+local build_path = os.getenv("BUILDDIR")
+package.cpath = build_path..'/test/sql-tap/?.so;'..build_path..'/test/sql-tap/?.dylib;'..package.cpath
+
+local test = require("sqltester")
+test:plan(43)
+
+local dec = require("decimal")
+local dec1 = dec.new("111")
+local dec2 = dec.new("55555")
+local dec3 = dec.new("3333")
```
Comment (Korablev77) [inline test/sql-tap/decimal.test.lua:11]: I'd also add dec4 with 0 value. There are several places where 0 is threatened as special case.

**Review by Korablev77 — APPROVED** [review]
Otherwise LGTM

File: src/box/sql/mem.c (hunk 4 above)
Comment (ImeevMA) [reply src/box/sql/mem.c]: Fixed.

File: src/box/sql/mem.c (hunk 5 above)
Comment (ImeevMA) [reply src/box/sql/mem.c]: Added.

File: src/box/sql/mem.c (hunk 6 above)
Comment (ImeevMA) [reply src/box/sql/mem.c:1070]: I think of it a bit later. @tsafin asked @sergepetrenko, who said this:

>I agree, it would be best to introduce decimal_to_double() to src/lib/core/decimal
>
>I can't think of an easy way here. The only one I see is construct the double value
>digit by digit. This mustn't be too different from decimal->string->double conversion. 

I plan to leave it as it is, for now. I still plan to fix this later.

File: test/sql-tap/decimal.test.lua (hunk 7 above)
Comment (ImeevMA) [reply test/sql-tap/decimal.test.lua:11]: Added.

**Review by VitaliyaIoffe — APPROVED** [review]
QA LGTM

**Review by kyukhin — APPROVED** [review]

**Comment by kyukhin** [comment]
Please, submit yet another PR w/ ChangeLog entry.
